### PR TITLE
[Merged by Bors] - feat(Algebra/Category/GroupCat/Abelian): prove AddCommGroupCat is AB5

### PR DESCRIPTION
**Pull request by Multramate**
This work was done during the 2023 Copenhagen masterclass on formalisation of condensed mathematics. Numerous participants contributed.

Co-authored-by: Moritz Firsching <moritz.firsching@gmail.com>
Co-authored-by: Nikolas Kuhn <nikolaskuhn@gmx.de>
Co-authored-by: Amelia Livingston <101damnations@github.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Category/GroupCat/Kernels.lean [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+variable {G H : AddCommGroupCat.{u}} (f : G ⟶ H)
+
+-- TODO: relocate theorem
+theorem range_le_ker_iff {G H I : AddCommGroupCat.{u}} {f : G ⟶ H} {g : H ⟶ I} :
+    f.range ≤ g.ker ↔ g.comp f = 0 :=
+  ⟨fun h => AddMonoidHom.ext fun x => h ⟨x, rfl⟩,
+    by rintro h _ ⟨x', rfl⟩; exact FunLike.congr_fun h x'⟩
+
+-- TODO: relocate theorem
```
Comment (alexjbest) [inline Mathlib/Algebra/Category/GroupCat/Kernels.lean:27]: Is this a to-do for now or for later? And the other ones nearby

Comment (Multramate) [reply Mathlib/Algebra/Category/GroupCat/Kernels.lean:27]: Whoops I completely forgot about these - I'll relocate them now

Comment (Multramate) [reply Mathlib/Algebra/Category/GroupCat/Kernels.lean]: Whoops I completely forgot about these - I've relocated them.

File: Mathlib/Algebra/Category/GroupCat/Abelian.lean [hunk 2]
```diff
@@ -47,18 +45,42 @@ def normalEpi (_ : Epi f) : NormalEpi f :=
...
+  add_comp := by exact Preadditive.add_comp
+  comp_add := by exact Preadditive.comp_add
+
+theorem exact_iff : Exact f g ↔ f.range = g.ker := by
+  rw [Abelian.exact_iff' f g (kernelIsLimit _) (cokernelIsColimit _)]
+  exact
+    ⟨fun h => ((range_le_ker_iff _ _).mpr h.left).antisymm ((ker_le_range_iff _ _).mpr h.right),
+      fun h => ⟨(range_le_ker_iff _ _).mp h.le, (ker_le_range_iff _ _).mp h.symm.le⟩⟩
+
+theorem exact_iff' : Exact f g ↔ f ≫ g = 0 ∧ g.ker ≤ f.range := by
```
Comment (TwoFX) [inline Mathlib/Algebra/Category/GroupCat/Abelian.lean:62]: You could think about turning this into a lemma `range_eq_ker_iff` that lives next to `range_le_ker_iff`.

**Review by TwoFX — CHANGES_REQUESTED** [review]

File: Mathlib/Algebra/Category/GroupCat/Abelian.lean (hunk 2 above)
Comment (Multramate) [reply Mathlib/Algebra/Category/GroupCat/Abelian.lean:62]: You mean `exact_iff`? This name is analogous to the one in `ModuleCat`, so if we'd have to change both.

File: Mathlib/Algebra/Category/GroupCat/Abelian.lean [hunk 3]
```diff
@@ -47,18 +45,44 @@ def normalEpi (_ : Epi f) : NormalEpi f :=
...
   normalMonoOfMono := normalMono
   normalEpiOfEpi := normalEpi
-  add_comp := by
-    intros
-    simp only [Preadditive.add_comp]
-  comp_add := by
-    intros
-    simp only [Preadditive.comp_add]
+  add_comp := by exact Preadditive.add_comp
+  comp_add := by exact Preadditive.comp_add
```
Comment (TwoFX) [inline Mathlib/Algebra/Category/GroupCat/Abelian.lean:54]: I think you can drop these entirely
```suggestion
```

File: Mathlib/Algebra/Category/GroupCat/Abelian.lean (hunk 2 above)
Comment (TwoFX) [reply Mathlib/Algebra/Category/GroupCat/Abelian.lean:64]: Sorry, I guess what I wanted to say is that the lemma `exact_iff'` doesn't really have any content. Indeed, if you add the missing
```lean
theorem comp_def : f ≫ g = g.comp f :=
  rfl
```
to `GroupCat/Basic.lean`, then `exact_iff'` is just
```lean
theorem exact_iff' : Exact f g ↔ f ≫ g = 0 ∧ g.ker ≤ f.range := by
  rw [exact_iff, le_antisymm_iff, AddMonoidHom.range_le_ker_iff, comp_def]
```
You can keep it with the shortened proof if you like, but you could also just inline it into the instance below.

**Review by TwoFX — APPROVED** [review]
Thanks!
bors r+

**Pull request closed without merging**